### PR TITLE
generalised the sync iterator so that it does not depend on the OrdVe…

### PR DESCRIPTION
**Pull request by kilohsakul**
Generalizing the synchronized_iterator so that it does not depend on OrdVector specifically but can take any container who has const_iterator.

### CODE REVIEW

**Comment by martinhruska** [comment]
PR looks good to me. I like the generalization of generalized iterator. Although I share some of youd scepticism on documentation it looks that you deleted the whole documentation folder by mistake. I suppose @tfiedor can provide some git magic to fix. This thing happens because once you call `make clean` it deletes the `doc/` folder and then if you call `git commit -am` you confirm deleting the folder . We need to fix it.

**Comment by kilohsakul** [comment]
> PR looks good to me. I like the generalization of generalized iterator. Although I share some of youd scepticism on documentation it looks that you deleted the whole documentation folder by mistake. I suppose @tfiedor can provide some git magic to fix. This thing happens because once you call `make clean` it deletes the `doc/` folder and then if you call `git commit -am` you confirm deleting the folder . We need to fix it.

Omg. Please help me resolve this.

**Comment by tfiedor** [comment]
> > PR looks good to me. I like the generalization of generalized iterator. Although I share some of youd scepticism on documentation it looks that you deleted the whole documentation folder by mistake. I suppose @tfiedor can provide some git magic to fix. This thing happens because once you call `make clean` it deletes the `doc/` folder and then if you call `git commit -am` you confirm deleting the folder . We need to fix it.
> 
> Omg. Please help me resolve this.

Assuming you have no changes in current repo (otherwise do `git stash` and after running the next two commands run `git stash pop`): 1. `git reset devel -- doc` should restore the files. 2. `git -am "doc: restore deleted files", 3. Push the changes.

This should do the trick. If you are afraid, I could maybe fix it for you.

**Comment by tfiedor** [comment]
Well, maybe it would be safer for me to fix it. Just ping me, that you will not commit anything to this branch, until I fix it. And I will fix it.

**Comment by kilohsakul** [comment]
Ok, thanks, not comitting anything from now on.

**Comment by tfiedor** [comment]
> Ok, thanks, not comitting anything from now on.

Done, fixed, your royal highness. Will review the code after lunch.

File: src/nfa/nfa-intersection.cc [hunk 1]
```diff
@@ -117,12 +117,14 @@ Nfa intersection(const Nfa& lhs, const Nfa& rhs, bool preserve_epsilon, ProductM
...
-        Mata::Util::SynchronizedUniverzalIterator<TransSymbolStates> sui(2);
-        sui.push_back(lhs.transitionrelation[pair_to_process.first]);
-        sui.push_back(rhs.transitionrelation[pair_to_process.second]);
-
-        while (sui.advance()) {
-            std::vector<TransitionList::const_iterator> moves = sui.get_current();
+        Mata::Util::SynchronizedUniversalIterator<Mata::Util::OrdVector<TransSymbolStates>::const_iterator> sync_iterator(2);
+        Mata::Util::push_back(sync_iterator,lhs.transitionrelation[pair_to_process.first]);
+        Mata::Util::push_back(sync_iterator,rhs.transitionrelation[pair_to_process.second]);
+        //sync_iterator.push_back(lhs.transitionrelation[pair_to_process.first]);
```
Comment (tfiedor) [inline src/nfa/nfa-intersection.cc:123]: I would advise not to leave commented out code in PRs. It will only confuse others and yourself in the future.

**Review by tfiedor — APPROVED** [review]
No major issues. I would advise to use `/**/` comments for documenting functions so it is uniform with the rest of the codes, but it doesn't pain me as much.

File: src/nfa/nfa-intersection.cc (hunk 1 above)
Comment (kilohsakul) [reply src/nfa/nfa-intersection.cc]: okok, thanks, done

**Pull request merged**